Lib/test/test_tabnanny.py

```diff
@@ -0,0 +1,353 @@
+
+        file_path = create_python_file(SOURCE_CODES["error_free"])
+        with open(file_path) as f:
+            tabnanny.process_tokens(tokenize.generate_tokens(f.readline))
+        self.assertFalse(MockNannyNag.called)
+
+    def test_with_errored_codes_samples(self):
+        """A python source code with whitespace related sampled problems."""
+
+        # "tab_space_errored_1": exicutes block under type == tokenize.INDENT
+        #                        at `tabnanny.process_tokens()`.
+        # "tab space_errored_2": exicutes block under
+        #                        `check_equal and type not in JUNK` condition at
+        #                        `tabnanny.process_tokens()`.
+
+        for key in ["tab_space_errored_1", "tab_space_errored_2"]:
+            file_path = create_python_file(SOURCE_CODES[key])
+            exception = tabnanny.NannyNag
+            with open(file_path) as f:
+                tokens = tokenize.generate_tokens(f.readline)
+                self.assertRaises(exception, tabnanny.process_tokens, tokens)
+
+class TestCommandLine(TestCase):
+    """Tests command line interface of `tabnanny`."""
+
+    def _validate_cmd(self, *args, return_code=0, stdout=b"", stderr=b""):
+        """Common function to assert the behaviour of command line interface.
+
+        Arguments:
+            * args        : Add arguments which will be directly feeded as the
+                            arguments to tabnanny command line.
+            * return_code : Expected return code from tabnanny command line
+                            interface.
+            * stdout      : Expected standard output text after executing
+                            tabnanny command line interface.
+            * stderr      : Expected standard output error text after executing
+                            tabnanny command line interface.
+        """
+        rc, out, err = script_helper.assert_python_ok('-m', 'tabnanny', *args)
+        self.assertEqual(out, stdout)
+        self.assertEqual(rc, return_code)
+        self.assertEqual(err, stderr)
+
+    def test_with_errored_file(self):
+        """Should displays error when errored python file is given."""
+        file_path = create_python_file(SOURCE_CODES["wrong_indented"])
+        stderr = b"%r: Indentation Error: " % (file_path,)
+        stderr += (b'unindent does not match any outer indentation level'
+                   b' (<tokenize>, line 3)')
+        self._validate_cmd(file_path, stderr=stderr)
+
+    def test_with_error_free_file(self):
+        """Should not display anything if python file is correctly indented."""
+        file_path = create_python_file(SOURCE_CODES["error_free"])
+        self._validate_cmd(file_path)
+
+    def test_command_usage(self):
+        """Should display usage on no arguments."""
+        path = findfile('tabnanny.py')
+        stderr = b'Usage: %s [-v] file_or_directory ...' % (path.encode('ascii'))
```

Comment:
What if the path is non-ASCII? Again, just check the general form of the output and don’t be too specific.